[Feat]: Mouse input support

pawaskar-shreya opened the pull request:
hey @mdales! 

I have been working on this issue for the past couple of days. I would apologize for the delay. 

So, I have tried to implement the mouse input support by creating  a mouse object that keeps track of its position, which buttons (left, right, or wheel button) are pressed, and a list of recent events like button clicks, movement, and scrolling. 

I have also tried to write the tests for the same. Could you please see my work and tell me if there's any more improvement needed?

mdales commented on src/mouse.ml:
```diff
@@ -0,0 +1,53 @@
+let clear t = {
+  t with
+  events = [];
+}
```
This just clears events, not other state, so I think the name should reflect that (similar to `add_event`).

pawaskar-shreya replied:
Yess, you are right! I'll make that change

pawaskar-shreya replied:
yess, this is also done! 

forgot to mention it in the below summary 😅😅

mdales commented on src/base.ml:
```diff
@@ -78,42 +79,44 @@ let run (title : string) (boot : boot_func option) (tick : tick_func) (s : Scree
         );
 
-        let updated_buffer = tick t s prev_buffer keys in
+        let updated_buffer = tick t s prev_buffer keys (Some mouse) in
```
Why is mouse optional? it doesn't look like it can ever be `None`?

pawaskar-shreya replied:
Noted! I'll get it fixed

pawaskar-shreya replied:
And also this!

Sorry to tag you at multiple places  😅

mdales commented on src/base.mli:
```diff
@@ -9,7 +9,7 @@ module PlatformKey : module type of Keysdl
 (** Function called once a start of run *)
 
-type tick_func = int -> Screen.t -> Framebuffer.t -> KeyCodeSet.t -> Framebuffer.t
+type tick_func = int -> Screen.t -> Framebuffer.t -> KeyCodeSet.t -> Mouse.t option -> Framebuffer.t
```
This isn't wrong, it's the obvious thing to do given what I've done until now, but I think it'd be great if you could work in this suggestion by @3pleX-dev in [this comment](https://github.com/claudiusFX/Claudius/issues/12#issuecomment-2732964406), and combine the KeyCodeSet.t and Mouse.t into a single structure here. This means we just have one slot in the function call (as we'll add controller support at some point too in #11), and most of the examples don't use input so they can just ignore one thing rather than ignoring two things.

pawaskar-shreya replied:
Yess yess, I've revisited the suggestion. I thought of actually first trying to implement Mouse.t seperately as I was just starting up with things and then update it later as being suggested. 

Now that it's been done, I'll work on trying to unify it.

mdales reviewed (changes requested):
Firstly, this is looking great: well structured, I like the division of SDL and generic mouse code, and you've added tests. A brilliant first contribution! 

Some requests/questions:

* If you could combine the mouse and keyboard inputs to `tick` into a single structure, that I think sets us up for other types of input, and stops that `tick` function interact from getting ever longer.

* The MLI files need some documentation in odoc style like with the other mli headers.

* I appreciate what you're doing with the event log, and that does allow for more flexibility on the caller, but because you rely on the caller to clear the event log, if someone isn't using the mouse, and the mouse exists, then that event list will just grown forever, taking up memory unnecessarily. I think if we want to have the event log, then the runloop must clear it on every tick, and the odoc comments in the MLI file must indicate that the user of Claudius must check events every tick or they'll be lost.

Hopefully those all make sense. Otherwise this PR is looking great, well done!

pawaskar-shreya commented:
hey @mdales! 👋👋

I have implemeted all the suggestions you gave. A quick summary of what all have been done:

- Grouped the keyboard and mouse inputs into a single record
- Fixed the documentation for the MLI files
- Updated the event log to clear on every tick

I have treid to implement all the things that you suggested. And it was fun doing it, but kindly tell me if there's anything more that we want to implement here 😊

Good-day!

mdales commented on src/base.ml:
```diff
@@ -78,42 +84,45 @@ let run (title : string) (boot : boot_func option) (tick : tick_func) (s : Scree
         );
 
-        let updated_buffer = tick t s prev_buffer keys in
+        let updated_buffer = tick t s prev_buffer input in
```
Did you test this? I was going to ask that you also as a follow on make a new example for Claudius-examples anyway, but my fear is you've not actually used this code to check it does what you think it does.

The event queue is cleared right before we pass mouse to the tick function, so the tick function will always see an empty event queue, right? So we keep track of the events, then throw them away right before we use them!

pawaskar-shreya replied:
I am glad that you liked the documentation and added tests! 😊

I had run tests, but should have taken more time to test it thoroughly. I will now go and see what mistakes I have done and report back to you.

mdales commented on test/dune:
```diff
@@ -1,3 +1,3 @@
- (names test_palette test_screen test_framebuffer test_primitives test_utils test_keysdl)
- (libraries claudius ounit2))
+ (names test_palette test_screen test_framebuffer test_primitives test_utils test_keysdl test_mousesdl)
+ (libraries claudius ounit2 tsdl))
```
I don't believe you need tsdl here - it's required by Claudius, so should automatically be found? Or did you have a problem with that?

pawaskar-shreya replied:
Yesss, I had seen it that Claudius already uses tsdl, But I was getting an error message at that time "Unbound module Tsdl" for my mousesdl.ml and in an attempt to fix it, I added tsdl in the dune file.  

Later I realized it was an opam switch issue.

But I will remove it my next commit.

mdales reviewed (changes requested):
This is looking really good, thank you for your effort, in particular the care you've put into documentation and unit tests.

However, there does appear to be a serious logic bug around clearing the events - can you look into this?

mdales commented:
I made a little test program, and I found an issue!

https://github.com/user-attachments/assets/840aa28b-8448-40b2-bc03-42a00bdc08cf

Here, as I often do, I'm using a scaling factor on the screen:

```ocaml
let () =
	Palette.generate_mac_palette () |>
	Screen.create 320 200 4 |>
	Base.run "Claudius Paint" (Some boot) tick
```

But the mouse coordinates we return are not similarly scaled! So in my code I had to manually adjust them:

```ocaml
		let x, y = Mouse.get_position inputs.mouse in
		let x = x / 4 and y = y / 4 in
```

Claudius should do this conversation for us!

pawaskar-shreya commented:
Good point, scaling was not previously taken care of. I will get it done today itself

mdales commented:
Due to the changes from #38, which changed the runloop, your branch now conflicts. You'll need to rebase and fix the conflicts - they look messy, but are not too bad: the run-loop structure has changed in base.ml, but if you replace keys with inputs again from the new code all should be well. You can see the result in the `mwd-fun-friday` branch where I merged your code with @Vanisha1606 's to let me test both branches.

pawaskar-shreya commented:
heyyy @mdales! 

I am done with the work on my local. Now trying to do the rebasing as now suggested. I am following your suggestion of making a backup branch for the current work from [https://github.com/claudiusFX/Claudius/pull/35](url)

pawaskar-shreya commented:
hello @mdales 

I have made the requested changes. Fixed the event queue issue as you had mentioned earlier. I have also tested it thoroughly this time:

![after fixing event loop clearing](https://github.com/user-attachments/assets/a0f94bb3-0a68-4912-80ca-b37732ed858b)


Also, the other issue that you had pointed out about scaling, I tried to resolve it by introduing a scale field to the Mouse.t 
And then I have updated respected functions for creting an event and updating position. This updated position is what we now get when we call Mouse.get_position

Here's the before and after:

Before scaling fix:
![Before scaling](https://github.com/user-attachments/assets/f8fa19a9-e19d-402d-90d5-777a1d569199)


After scaling fix:
![After Scaling](https://github.com/user-attachments/assets/a582a69a-3cdc-4cbb-abcd-a6fc5c9bc783)


I have also incorporated the reabase method as you suggested. The merge conflicts were resolved and git log shows all the commits of other people also along with mine. 

Please check once and let me know if I have missed something or if there is any change that needs to be done.

Thankyou!

pawaskar-shreya commented:
hey @mdales! 

There seems to be some serious performance issue arising after the recent commits! I had tested for the changes I made in correspondence to the scaling issue. That issue seems to be solved.

So, until you review my code, I tried to test the recent changes as you did on my local machine. It seems to be working correctly logically. But there's a serious overhead between the time when we draw with our mouse and when the drawing acutally appers on the screen. 

https://github.com/user-attachments/assets/9d3ba43c-b319-429c-9e2b-6273ad486036


Also, when I try to record my screen with the snipping tool available on windows, i cannot seem to capture the first letter drawn. Tried it multiple times, the same issue persists. 

https://github.com/user-attachments/assets/b4aaeaa3-c835-4e7b-878c-c6d835942a82


I can't tell if it is arising because I am running running Claudius on ubuntu throught WSL or is it some different issue. 

I will try to test without the scaling changes and report back to you. Please suggest me what you think here might be causing this issue.

pawaskar-shreya commented:
update: 

I have done the testing without the scaling code and harcoded scaling just like you did in paint.ml, but the issue still persists. This makes me want to think in the direction that maybe it might to due to me running the code on WSL and some resource optimization taking place. 

Please correct me if I am thinking in wrong direction.

mdales commented:
Yeah, I'd be surprised if it's the scaling or anything here, it just looks like Claudius is being very slow - which is both good and bad news. Good in that I don't think it should stop your PR, but bad as clearly that's unusable performance. Can you raise an issue for the poor performance, then we start a discussion there on how to measure it and work out if we can find a cause. Whilst I have used WSL with Claudius in the past, it has been a good while since I did so. This is a place where #30 with an FPS counter would be very useful.

mdales commented on src/mouse.mli:
```diff
@@ -0,0 +1,41 @@
+val create : int -> t
+(** [create scale] initializes a new mouse state with a given scale factor. *)
+
+val clear_event : t -> t
```
This should be plural - sorry to be picky, but once you and an API it's very hard to change after people start using it.

pawaskar-shreya replied:
Yess yess, I'll go and fix it. Just a minor change there and it's better that we fix it now. It will save us from future inconveniences.

mdales commented on test/test_mousesdl.ml:
```diff
@@ -0,0 +1,104 @@
+  Sdl.Event.set event Sdl.Event.mouse_motion_y 250;
+  
+  let mouse = Mousesdl.handle_mouse_motion_event event mouse in
+  assert_equal (150, 250) (Mouse.get_position mouse)
```
Can we have a version of this where scale isn't 1?

pawaskar-shreya replied:
Okayy I'll try to make this change, that day just for making it accept scale I updated it to 1. But I'll definitely try this.

mdales commented on src/mouse.ml:
```diff
@@ -0,0 +1,63 @@
+  events = [];
+  position = (0, 0);
+  buttons = [(Left, false); (Middle, false); (Right, false)];
+  scale;
```
Given scale is int perhaps we should sanity check it isn't 0 or negative? (and then have a test for those in the tests)

pawaskar-shreya replied:
Noted!

mdales reviewed (changes requested):
Looking great - some very picky comments from me, but particularly the API naming I'd like to get fixed as that sort of thing is hard to correct later

pawaskar-shreya commented:
`Can you raise an issue for the poor performance, then we start a discussion there on how to measure it and work out if we can find a cause`

Yess, sure, I'll do that, there was also a segmentation fault issue arising. I found a temporary fix for that. I have the screenshot of the issue with me on my laptop. I'll add that also, may be it's linked to the same issue. 

I'm currently outside for today and don't have my laptop with me. But I'll do the changes and putting up the issue as soon as I get back home.

mdales commented:
> > Can you raise an issue for the poor performance, then we start a discussion there on how to measure it and work out if we can find a cause
> 
> Yess, sure, I'll do that, there was also a segmentation fault issue arising. I found a temporary fix for that. I have the screenshot of the issue with me on my laptop. I'll add that also, may be it's linked to the same issue.
> 
> I'm currently outside for today and don't have my laptop with me. But I'll do the changes and putting up the issue as soon as I get back home.
> 
> Apologies for the delay.

Thanks - and no need to apologise, take your time!

mdales commented:
Just took this for a test, and the scale changes are working great. There's still a couple of requests I made on the last review that need done, but otherwise looking ready to merge!

pawaskar-shreya commented:
hey @mdales done with adding sanity check for scaling and updating the tests for it!

mdales reviewed (approved):
Looks great, many thanks for this, this opens up a whole other avenue of ways to use Claudius!
